libs/utils/energy_model.py

```diff
@@ -0,0 +1,572 @@
+
+    def _cpus_with_capacity(self, cap):
+        """
+        Helper method to find the CPUs whose max capacity equals cap
+        """
+        return [c for c in self.cpus
+                if self.cpu_nodes[c].max_capacity == cap]
+
+    @property
+    @memoized
+    def biggest_cpus(self):
+        """
+        The CPUs with the highest compute capacity at their highest frequency
+        """
+        return self._cpus_with_capacity(self.capacity_scale)
+
+    @property
+    @memoized
+    def littlest_cpus(self):
+        """
+        The CPUs with the lowest compute capacity at their highest frequency
+        """
+        min_cap = min(n.max_capacity for n in self.cpu_nodes)
+        return self._cpus_with_capacity(min_cap)
+
+    @property
+    @memoized
+    def is_heterogeneous(self):
+        """
+        True iff CPUs do not all have the same efficiency and OPP range
+        """
+        states = self.cpu_nodes[0].active_states
+        return any(c.active_states != states for c in self.cpu_nodes[1:])
+
+    def _guess_idle_states(self, cpus_active):
+        def find_deepest(pd):
+            if not any(cpus_active[c] for c in pd.cpus):
+                if pd.parent:
+                    parent_state = find_deepest(pd.parent)
+                    if parent_state:
+                        return parent_state
+                return pd.idle_states[-1] if len(pd.idle_states) else None
+            return None
+
+        return [find_deepest(pd) for pd in self.cpu_pds]
+
+    def guess_idle_states(self, cpus_active):
+        """Pessimistically guess the idle states that each CPU may enter
+
+        If a CPU has any tasks it is estimated that it may only enter its
+        shallowest idle state in between task activations. If all the CPUs
+        within a power domain have no tasks, they will all be judged able to
+        enter that domain's deepest idle state. If any CPU in a domain has work,
+        no CPUs in that domain are assumed to enter any domain shared state.
+
+        e.g. Consider a system with
+
+        - two power domains PD0 and PD1
+
+        - 4 CPUs, with CPUs [0, 1] in PD0 and CPUs [2, 3] in PD1
+
+        - 4 idle states: "WFI", "cpu-sleep", "cluster-sleep-0" and "cluster-sleep-1"
+
+        Then here are some example inputs and outputs:
+
+        ::
+
+          # All CPUs idle:
+          [0, 0, 0, 0] -> ["cluster-sleep-0", "cluster-sleep-0",
+                           "cluster-sleep-0", "cluster-sleep-0"]
+
+          # All CPUs have work
+          [1, 1, 1, 1] -> ["WFI","WFI","WFI", "WFI"]
+
+          # One power domain active, the other idle
+          [0, 0, 1, 1] -> ["cluster-sleep-1", "cluster-sleep-1", "WFI","WFI"]
+
+          # One CPU active.
+          # Note that CPU 2 has no work but is assumed to never be able to enter
+          # any "cluster" state.
+          [0, 0, 0, 1] -> ["cluster-sleep-1", "cluster-sleep-1",
+                           "cpu-sleep","WFI"]
```

Comment:
No:

> If a CPU has any tasks it is estimated that it may only enter its
        shallowest idle state in between task activations

And in line 346 all the CPUs have tasks so can only enter WFI